File: solution/math/5618/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/1df904530afc47edbe9a9983b130ce19
#include<bits/stdc++.h>

using namespace std;

int gcd(int a, int b) {
    if(b == 0) return a;
    return gcd(b, a%b);
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n; cin >> n;
    int a, b; cin >> a >> b;
    a = gcd(a, b);
    if(n == 3) {
        int c; cin >> c;
        a = gcd(a, c);
    }

    vector<int> ans;
    for(int i=1;i*i<=a;i++){
        if(a % i != 0) continue;
        ans.push_back(i);
        if(i * i != a) ans.push_back(a / i);
    }
    sort(ans.begin(), ans.end());
    for(int i=0;i<(int)ans.size();i++){
        cout << ans[i] << '\n';
    }

    return 0;
}

File: solution/math/5618/main.py
# Authored by : gusdn3477
# Co-authored by : tony9402
# Link : http://boj.kr/aef57ade1e5c4c6e90f08a159fe96ca2
import sys

def input():
    return sys.stdin.readline().rstrip()

def GCD(x,y):
    if y == 0:
        return x
    else:
        return GCD(y, x%y)

n = int(input())
arr = list(map(int, input().split()))
outputs = list()

gcd = arr[0]
for i in range(1, n):
    gcd = GCD(gcd, arr[i])

x = 1
while x * x <= gcd:
    if gcd % x == 0:
        outputs.append(x)
        if x * x != gcd:
            outputs.append(gcd // x)
    x += 1

outputs.sort()
print(*outputs)
